From b865fa9690d0ab0889a593f481ec6775c075ae05 Mon Sep 17 00:00:00 2001
From: Alexei Chetroi <Alexei.Chetroi@outlook.com>
Date: Wed, 27 Dec 2023 12:20:48 -0500
Subject: [PATCH] Custom commands for level cluster

---
 .../plugin/level-control/level-control.c      | 30 +++++++++++++++----
 1 file changed, 24 insertions(+), 6 deletions(-)

diff --git a/protocol/zigbee/app/framework/plugin/level-control/level-control.c b/protocol/zigbee/app/framework/plugin/level-control/level-control.c
index 9cef92991..c2acb43ca 100644
--- a/protocol/zigbee/app/framework/plugin/level-control/level-control.c
+++ b/protocol/zigbee/app/framework/plugin/level-control/level-control.c
@@ -550,7 +550,14 @@ static void moveToLevelHandler(uint8_t commandId,
                                uint8_t optionOverride,
                                uint16_t storedLevel)
 {
-  uint8_t endpoint = emberAfCurrentEndpoint();
+  bool sendResponse = true;
+  uint8_t endpoint;
+  if (emberAfCurrentCommand() == NULL) {
+      endpoint = emberAfPrimaryEndpoint();
+      sendResponse = false;
+  } else {
+      endpoint = emberAfCurrentEndpoint();
+  }
   EmberAfLevelControlState *state = getState(endpoint);
   EmberAfStatus status;
   uint8_t currentLevel;
@@ -670,7 +677,8 @@ static void moveToLevelHandler(uint8_t commandId,
 #endif
 
   send_default_response:
-  if (emberAfCurrentCommand()->apsFrame->clusterId
+  if (sendResponse
+      && emberAfCurrentCommand()->apsFrame->clusterId
       == ZCL_LEVEL_CONTROL_CLUSTER_ID) {
     emberAfSendImmediateDefaultResponse(status);
   }
@@ -678,7 +686,14 @@ static void moveToLevelHandler(uint8_t commandId,
 
 static void moveHandler(uint8_t commandId, uint8_t moveMode, uint8_t rate, uint8_t optionMask, uint8_t optionOverride)
 {
-  uint8_t endpoint = emberAfCurrentEndpoint();
+  bool sendResponse = true;
+  uint8_t endpoint;
+  if (emberAfCurrentCommand() == NULL) {
+      endpoint = emberAfPrimaryEndpoint();
+      sendResponse = false;
+  } else {
+      endpoint = emberAfCurrentEndpoint();
+  }
   EmberAfLevelControlState *state = getState(endpoint);
   EmberAfStatus status;
   uint8_t currentLevel;
@@ -781,7 +796,7 @@ static void moveHandler(uint8_t commandId, uint8_t moveMode, uint8_t rate, uint8
   status = EMBER_ZCL_STATUS_SUCCESS;
 
   send_default_response:
-  emberAfSendImmediateDefaultResponse(status);
+  if (sendResponse) emberAfSendImmediateDefaultResponse(status);
 }
 
 static void stepHandler(uint8_t commandId,
@@ -984,11 +999,14 @@ void emberAfOnOffClusterLevelControlEffectCallback(uint8_t endpoint,
 
   if (newValue) {
     // If newValue is ZCL_ON_COMMAND_ID...
-    // "Set CurrentLevel to minimum level allowed for the device."
+    // "Set CurrentLevel to minimum level allowed for the device",
+    // unless we just make it instantaneous, depending on OnOff transition time
+    bool instanteneous = currentOnOffTransitionTime == 0x000
+                         || currentOnOffTransitionTime == 0xFFFF;
     status = emberAfWriteServerAttribute(endpoint,
                                          ZCL_LEVEL_CONTROL_CLUSTER_ID,
                                          ZCL_CURRENT_LEVEL_ATTRIBUTE_ID,
-                                         (uint8_t *)&minimumLevelAllowedForTheDevice,
+                                         (uint8_t *) (instanteneous) ? &resolvedLevel : &minimumLevelAllowedForTheDevice,
                                          ZCL_INT8U_ATTRIBUTE_TYPE);
     if (status != EMBER_ZCL_STATUS_SUCCESS) {
       emberAfLevelControlClusterPrintln("ERR: reading current level %x", status);
-- 
2.40.1.windows.1

